### DISCUSSION TITLE
How to deal with non-standard `package_data`?

### DISCUSSION THREAD

**Discussion by brianthelion** [pyscaffold/pyscaffold#507] [Q&A] [answered]
[This `setuptools` bug](https://github.com/pypa/setuptools/issues/1806) means that recursive globbing of `options.package_data` a-la `* = **/*.yaml` doesn't work.

Furthermore, with PyScaffold, the following `setup.py` recipe doesn't seem to have any effect:

```python
...
if __name__ == "__main__":
    try:
        setup(use_scm_version={"version_scheme": "no-guess-dev"},
              package_data={"": FILES_I_CARE_ABOUT})
    except:  # noqa
        print(
            "\n\nAn error occurred while building the project, "
            "please ensure you have the most updated version of setuptools, "
            "setuptools_scm and wheel with:\n"
            "   pip install -U setuptools setuptools_scm wheel\n\n"
        )
        raise
```
I have verified this by manually unzipping the `.whl` after `python setup.py bdist_wheel`.

PyScaffold is v4.0.2 and setuptools is v58.1.0.

Please advise.
Thanks!

**Comment by abravalheri** [comment]
Hi @brianthelion, please find some considerations bellow:

- Could this issue be related to https://github.com/pypa/setuptools/issues/1461 and https://github.com/pyscaffold/pyscaffold/issues/278? (By default `setup.cfg` comes with `include_package_data = True`, right?)
- Having this bug in mind, do you think it is possible to achieve what you are trying to achieve with a `include_package_data = True` and `exclude_package_data` ([ref1](https://setuptools.pypa.io/en/latest/userguide/datafiles.html), [ref2](https://pyscaffold.org/en/stable/features.html#configuration-packaging-distribution))?
- Although `setuptools` documentation seems to indicate otherwise, I would also try to add a `__init__.py` file under the directories with the data files... Just to see if it works...

Finally, if those comments are not helpful, could you please share a minimal repository that replicates your problem? I might have more insights... (it can be completely made up files, you don't have to disclose your data or python code :P)

**Reply by brianthelion** [reply]
@abravalheri I'll pull together a minimal example shortly.

**Reply by brianthelion** [reply]
@abravalheri [This new `setuptools` issue](https://github.com/pypa/setuptools/issues/2805) may be a confound. I'm investigating further.

**Reply by brianthelion** [reply]
Minimal Dockerfile:

```Dockerfile
FROM python:3.8-slim
RUN apt-get update
RUN apt-get install -y git tree unzip
RUN git config --global user.email "you@example.com"
RUN git config --global user.name "Your Name"
RUN pip install --upgrade pip setuptools wheel
RUN pip install pyscaffold
RUN putup my-package

WORKDIR my-package
RUN mkdir src/foo
RUN touch src/foo/foo1.yaml
RUN mkdir src/bar
RUN touch src/bar/bar1.yaml
RUN mkdir src/zha.baz
RUN touch src/zha.baz/zhab.yaml
RUN echo src/foo/foo1.yaml >> files_i_care_about
RUN echo src/bar/bar1.yaml >> files_i_care_about
RUN echo src/zha.baz/zhab.yaml >> files_i_care_about

RUN echo 'import setuptools; import os; setuptools.setup(package_data={"": open("files_i_care_about").read().split(os.linesep)})' > setup.py
RUN tree
RUN python setup.py bdist_wheel
RUN unzip -l dist/*.whl
```

**Reply by FlorianWilhelm** [reply]
Hi @brianthelion, is this an example of how you built your package? In practice I assume you would rather copy in an actual Git repository containing the structure you are setting here up with `RUN mkdir/touch` etc. You have to keep in mind that PyScaffold is using `setuptools-scm` to track package data. What you would need to do in your example is to `git add` all `yaml` files (e.g. use `find ... -exec git ... ` for this) and commit them. Then it will just work.

**Reply by abravalheri** [reply]
@FlorianWilhelm have a very good point: I forgot to mention that you need to add the files to git so `setuptools-scm` can find them 😝

But yes, I can see here that `zha.baz/zhab.yaml` is not included in the wheel file.

My guess for what is happening is that you are unknowingly taking advantage of [PEP 420](https://www.python.org/dev/peps/pep-0420/) for the directories `src/bar` and `src/foo`, but since `zha.baz` is not a valid package name, it is not being included in the distribution.
(Please notice both `setup.cfg` and `docs/conf.py` templates from PyScaffold are tuned for PEP 420).

A way of testing this guess is:
```bash
sed -i 's/find_namespace:/find:/g' setup.cfg
rm -rf dist build src/*.egg-info
git add . && git commit -m "Disable PEP 420-aware packing"
python setup.py bdist_wheel
unzip -l dist/*.whl
```
You should not see any data file/folder
Then, reversing the process:
```bash
sed -i 's/find:/find_namespace:/g' setup.cfg
rm -rf dist build src/*.egg-info
git add . && git commit -m "Re-enable PEP 420-aware packing"
python setup.py bdist_wheel
unzip -l dist/*.whl
```
Then you can see both `foo` and `bar` directories in the wheel (but not `zha.baz` because of this name is not a valid module name for Python)

If you disable PEP 420, my guess is that the files are not included because of the same problem described in pypa/setuptools#1461.

**Reply by abravalheri** [reply]
If you are in a rush and need these things urgently, I recommend adding the files inside a folder inside your package like we describe in [the PACKAGE AND FILES DATA section of our docs](https://pyscaffold.org/en/stable/features.html#configuration-packaging-distribution). The advantage here is that you can use `importlib.resources` or `pkgutil` to read them easily and not having to worry about where your package is being placed (e.g. virtual envs). The disadvantage is that you (*might, I am not sure*)  have to make your directory names valid Python module names and add `__init__.py` files to them...

---
**Update**: Things seem to work fine when simply moving the files/directories to a folder inside `src/my_package` (see [ref](https://github.com/pypa/setuptools/issues/2805#issuecomment-932732542)), without requiring change the names of the folders or adding `__init__.py` files. However this is trickier to get right, so I still recommend the approach described above.

**Reply by abravalheri** [reply]
I think [my answers in the setuptools issue](https://github.com/pypa/setuptools/issues/2805#issuecomment-932732542) summarises the problem:

According to the [latest documentation](https://setuptools.pypa.io/en/latest/userguide/datafiles.html#non-package-data-files), `setuptools` [no longer supports](https://packaging.python.org/guides/distributing-packages-using-setuptools/?highlight=data%20files#package-data) data files outside of the package directory.

**Comment by abravalheri** [comment, marked as answer]
Just as a reference for users that might find a similar problem in the future:

- A deep investigation of why this problem happens can be found [here](https://github.com/pyscaffold/pyscaffold/discussions/507#discussioncomment-1417947) and in the linked pypa/setuptools#2805 issue.
- `setuptools` [no longer supports](https://setuptools.pypa.io/en/latest/userguide/datafiles.html#non-package-data-files) data files outside the package folder.
- The recommended solution is to put data files inside the package folder as described in in [the PACKAGE AND FILES DATA section of our docs](https://pyscaffold.org/en/latest/features.html#configuration-packaging-distribution).

**Reply by brianthelion** [reply]
@abravalheri Appreciate the diligence here. Can you include a revised version of the Dockerfile, above, to clarify the full solution? Thanks!

**Reply by abravalheri** [reply]
Hi @brianthelion, I will leave the Dockerfile as "an exercise for the reader" 😝, but the sequence of commands are:

```shell
# Create package
$ putup my-package
$ cd my-package

#  Data files should go inside a subdirectory of the package itself according to latest setuptools docs.
$ mkdir -p src/my_package/data/{foo,bar,zha.baz}
$ touch src/my_package/data/{foo/foo1.yaml,bar/bar1.yaml,zha.baz/zhab.yaml}

# Since `foo` and `bar` are valid identifiers, data files are automatically included due to `include_package_data = True`
# We just need to make them actual sub-packages
$ touch src/my_package/data/{foo,bar}/__init__.py

# `zha.baz` is not a valid identifier, so we need to add it to `package_data`
$ sed -i -E '/^\[options\.packages\.find\]$/i [options.package_data]\n* = data/zha.baz/*.yaml\n\n' setup.cfg
# ^------- The GNU version of `sed` (not sure if it will work on the BSD version).

# Add files to git, so `setuptools-scm` add them to manifest automatically
$ git add . && git commit -m "Add data files"

# Build wheel
$ rm -rf dist build && python setup.py bdist_wheel
#                      ^---------------------------- `tox -e build` can also be used here

$ unzip -l dist/*.whl
Archive:  dist/my_package-0.0.post1.dev2+gb768f1c-py3-none-any.whl
  Length      Date    Time    Name
---------  ---------- -----   ----
      581  2021-10-18 09:41   my_package/__init__.py
     4247  2021-10-18 09:41   my_package/skeleton.py
        0  2021-10-18 09:46   my_package/data/bar/__init__.py
        0  2021-10-18 09:47   my_package/data/bar/bar1.yaml
        0  2021-10-18 09:46   my_package/data/foo/__init__.py
        0  2021-10-18 09:47   my_package/data/foo/foo1.yaml
        0  2021-10-18 09:47   my_package/data/zha.baz/zhab.yaml
     1086  2021-10-18 10:05   my_package-0.0.post1.dev2+gb768f1c.dist-info/LICENSE.txt
     2408  2021-10-18 10:05   my_package-0.0.post1.dev2+gb768f1c.dist-info/METADATA
       92  2021-10-18 10:05   my_package-0.0.post1.dev2+gb768f1c.dist-info/WHEEL
       11  2021-10-18 10:05   my_package-0.0.post1.dev2+gb768f1c.dist-info/top_level.txt
     1075  2021-10-18 10:05   my_package-0.0.post1.dev2+gb768f1c.dist-info/RECORD
---------                     -------
     9500                     12 files
```  

### References:
- https://setuptools.pypa.io/en/latest/userguide/datafiles.html
- https://setuptools.pypa.io/en/latest/userguide/declarative_config.html
